src/panels/lovelace/common/generate-lovelace-config.ts

```diff
@@ -41,47 +39,56 @@ const HIDE_DOMAIN = new Set([
 
 const HIDE_PLATFORM = new Set(["mobile_app"]);
 
-interface SplittedByAreas {
-  areasWithEntities: Array<[AreaRegistryEntry, HassEntity[]]>;
+interface SplittedByAreaDevice {
+  areasWithEntities: { [areaId: string]: HassEntity[] };
+  devicesWithEntities: { [deviceId: string]: HassEntity[] };
   otherEntities: HassEntities;
 }
 
-const splitByAreas = (
-  areaEntries: AreaRegistryEntry[],
-  deviceEntries: DeviceRegistryEntry[],
-  entityEntries: EntityRegistryEntry[],
+const splitByAreaDevice = (
+  deviceEntries: HomeAssistant["devices"],
+  entityEntries: HomeAssistant["entities"],
   entities: HassEntities
-): SplittedByAreas => {
+): SplittedByAreaDevice => {
   const allEntities = { ...entities };
-  const areasWithEntities: SplittedByAreas["areasWithEntities"] = [];
-
-  for (const area of areaEntries) {
-    const areaEntities: HassEntity[] = [];
-    const areaDevices = new Set(
-      deviceEntries
-        .filter((device) => device.area_id === area.area_id)
-        .map((device) => device.id)
-    );
-    for (const entity of entityEntries) {
-      if (
-        ((areaDevices.has(
-          // @ts-ignore
-          entity.device_id
-        ) &&
-          !entity.area_id) ||
-          entity.area_id === area.area_id) &&
-        entity.entity_id in allEntities
-      ) {
-        areaEntities.push(allEntities[entity.entity_id]);
-        delete allEntities[entity.entity_id];
+  const areasWithEntities: SplittedByAreaDevice["areasWithEntities"] = {};
+  const devicesWithEntities: SplittedByAreaDevice["devicesWithEntities"] = {};
+
+  const areaDevices = new Set(
+    Object.values(deviceEntries)
+      .filter((device) => device.area_id)
+      .map((device) => device.id)
+  );
+  for (const entity of Object.values(entityEntries)) {
+    if (
+      (entity.area_id ||
+        (entity.device_id && areaDevices.has(entity.device_id))) &&
+      entity.entity_id in allEntities
+    ) {
+      const areaId =
+        entity.area_id || deviceEntries[entity.device_id!].area_id!;
+      if (!(areaId in areasWithEntities)) {
+        areasWithEntities[areaId] = [];
       }
-    }
-    if (areaEntities.length > 0) {
-      areasWithEntities.push([area, areaEntities]);
+      areasWithEntities[areaId].push(allEntities[entity.entity_id]);
+      delete allEntities[entity.entity_id];
+    } else if (entity.device_id && entity.entity_id in allEntities) {
+      if (!(entity.device_id in devicesWithEntities)) {
+        devicesWithEntities[entity.device_id] = [];
+      }
+      devicesWithEntities[entity.device_id].push(allEntities[entity.entity_id]);
+      delete allEntities[entity.entity_id];
     }
   }
+  Object.entries(devicesWithEntities).forEach(([deviceId, deviceEntities]) => {
+    if (deviceEntities.length === 1) {
+      allEntities[deviceEntities[0].entity_id] = deviceEntities[0];
+      delete devicesWithEntities[deviceId];
+    }
+  });
```

Comment:
We already have for...of in this function. Or we can convert other `for...of` to `forEach` for consistency.